tck/src/main/java/org/reactivestreams/tck/TestEnvironment.java

```diff
@@ -886,34 +884,39 @@ public Promise(TestEnvironment env) {
     }
 
     private ArrayBlockingQueue<T> abq = new ArrayBlockingQueue<T>(1);
-    private volatile T _value = null;
+    private AtomicReference<T> _value = new AtomicReference<T>();
 
     public T value() {
-      if (isCompleted()) {
-        return _value;
+      final T value = _value.get();
+      if (value != null) {
+        return value;
       } else {
         env.flop("Cannot access promise value before completion");
         return null;
       }
     }
 
     public boolean isCompleted() {
-      return _value != null;
+      return _value.get() != null;
     }
 
     /**
      * Allows using expectCompletion to await for completion of the value and complete it _then_
      */
     public void complete(T value) {
-      abq.add(value);
+      if (_value.compareAndSet(null, value)) {
+        // we add the value to the queue such to wake up any expectCompletion which was triggered before complete() was called
+        abq.add(value);
```

Comment:
Don't we want to track 2..N completions? (by publishing to the abq no matter the current status?)

Reply:
Ah, good point, indeed multiple completions are forbidden. In the way we use this Promise it's best to flop right away here I think